app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt

```diff
@@ -0,0 +1,372 @@
+     * app is run on a given server instance. So, after the first authentication, they are
+     * saved in SharedPreferences and every subsequent run they are simply fetched from there.
+     */
+    private fun onButtonClick() {
+
+        loginButton.isEnabled = false
+
+        domain = validateDomain(domainEditText.text.toString())
+
+        val callback = object : Callback<AppCredentials> {
+            override fun onResponse(call: Call<AppCredentials>,
+                                    response: Response<AppCredentials>) {
+                if (!response.isSuccessful) {
+                    loginButton.isEnabled = true
+                    domainEditText.error = getString(R.string.error_failed_app_registration)
+                    Log.e(TAG, "App authentication failed. " + response.message())
+                    return
+                }
+                val credentials = response.body()
+                clientId = credentials!!.clientId
+                clientSecret = credentials.clientSecret
+
+                redirectUserToAuthorizeAndLogin(domainEditText)
+            }
+
+            override fun onFailure(call: Call<AppCredentials>, t: Throwable) {
+                loginButton.isEnabled = true
+                domainEditText.error = getString(R.string.error_failed_app_registration)
+                Log.e(TAG, Log.getStackTraceString(t))
+            }
+        }
+
+        try {
+            getApiFor(domain)
+                    .authenticateApp(getString(R.string.app_name), oauthRedirectUri,
+                            OAUTH_SCOPES, getString(R.string.app_website))
+                    .enqueue(callback)
+            setLoading(true)
+        } catch (e: IllegalArgumentException) {
+            setLoading(false)
+            domainEditText.error = getString(R.string.error_invalid_domain)
+        }
+
+    }
+
+    private fun redirectUserToAuthorizeAndLogin(editText: EditText) {
+        /* To authorize this app and log in it's necessary to redirect to the domain given,
+         * activity_login there, and the server will redirect back to the app with its response. */
+        val endpoint = MastodonApi.ENDPOINT_AUTHORIZE
+        val redirectUri = oauthRedirectUri
+        val parameters = HashMap<String, String>()
+        parameters["client_id"] = clientId!!
+        parameters["redirect_uri"] = redirectUri
+        parameters["response_type"] = "code"
+        parameters["scope"] = OAUTH_SCOPES
+        val url = "https://" + domain + endpoint + "?" + toQueryString(parameters)
+        val uri = Uri.parse(url)
+        if (!openInCustomTab(uri, this)) {
+            val viewIntent = Intent(Intent.ACTION_VIEW, uri)
+            if (viewIntent.resolveActivity(packageManager) != null) {
+                startActivity(viewIntent)
+            } else {
+                editText.error = getString(R.string.error_no_web_browser_found)
+            }
+        }
```

Comment:
We have an issue here. If user closes custom tab we will be loading *forever*. If we put `setLoading(false)` here then we should `setLoading(true)` again in another place. Another option would be detecting when the custom tab is closed but I'm not sure it's possible (other then `onResume()`)